BJS2-78038 system comment

Fagan89 opened the pull request:
Система комментариев.

Commit:
BJS2-78038
Создал классы (!!!не рабочий код)

Commit:
BJS2-78038
Система комментариев
(тест не доделаны)

tanufa25 commented on src/main/java/faang/school/postservice/controller/comment/CommentController.java:
```diff
@@ -0,0 +1,49 @@
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/v1/")
```
После v1 слеш вроде лишний. Или в начале каждого эндпоинта нужно убрать, а то дублируется.

tanufa25 commented on src/main/java/faang/school/postservice/controller/comment/CommentController.java:
```diff
@@ -0,0 +1,49 @@
+        return ResponseEntity.ok(commentDtoResponse);
+    }
+
+    @PutMapping("/comments/{commentId}")
```
{commentId} - эта PathVariable не используется?

tanufa25 commented on src/main/java/faang/school/postservice/controller/comment/CommentController.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @GetMapping("/comments")
+    public ResponseEntity<List<CommentDtoResponse>> getAllComment(long postId) {
```
Включу душнилу и скажу, что мне буквы s в конце названия метода не хватает)

tanufa25 commented on src/main/java/faang/school/postservice/controller/comment/CommentController.java:
```diff
@@ -0,0 +1,49 @@
+
+    @GetMapping("/comments")
+    public ResponseEntity<List<CommentDtoResponse>> getAllComment(long postId) {
+        List<CommentDtoResponse> commentDtoResponseList = commentServiceF.getAllComment(postId);
```
getAllComments хочется

tanufa25 commented on src/main/java/faang/school/postservice/dto/comment/CommentCreateDto.java:
```diff
@@ -0,0 +1,21 @@
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
```
В моих пр Дима всегда на Data ругается, говорит оставить только то, что нужно

Fagan89 replied:
В DTO разрешил

tanufa25 commented on src/main/java/faang/school/postservice/dto/comment/CommentDtoResponse.java:
```diff
@@ -0,0 +1,20 @@
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
```
.

tanufa25 commented on src/main/java/faang/school/postservice/dto/comment/CommentUpdateDto.java:
```diff
@@ -0,0 +1,20 @@
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
```
.

tanufa25 commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,57 @@
+    public Comment createComment(Comment comment) {
+        commentValidation.validateLengthContentComment(comment);
+        commentValidation.validateAuthorExists(comment);
+        commentValidation.validatePostExists(comment);
```
Может, есть смысл эти 3 валидации вынести в отдельный метод? А то в update тоже эти 3 валидации дёргаются.

tanufa25 commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,57 @@
+        return commentRepository.save(comment);
+    }
+
+    public List<Comment> getAllComment(long postId) {
```
getAllComments

tanufa25 commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    public List<Comment> getAllComment(long postId) {
+        if (!postRepository.existsById(postId)) {
+            throw new EntityNotFoundException(("the post does not exists %d".formatted(postId)));
+        }
```
Может, это тоже в отдельный метод валидации? И перед выбрасыванием ошибки лучше сделать лог о ней log.error("message text")

tanufa25 commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    public void deleteComment(long commentId) {
+        if (!commentRepository.existsById(commentId)) {
+            throw new EntityNotFoundException(("the comment does not exists %d".formatted(commentId)));
+        }
```
И тут, возможно, лучше в отдельный метод и лог добавить

tanufa25 commented on src/main/java/faang/school/postservice/service/comment/CommentServiceFacade.java:
```diff
@@ -0,0 +1,44 @@
+        return mapperComment.fromEntityToDto(commentUpdate);
+    }
+
+    public List<CommentDtoResponse> getAllComment(long postId) {
```
s

tanufa25 commented on src/main/java/faang/school/postservice/service/comment/CommentServiceFacade.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    public List<CommentDtoResponse> getAllComment(long postId) {
+        List<Comment> comments = commentService.getAllComment(postId);
```
s

tanufa25 commented on src/main/java/faang/school/postservice/validation/comment/CommentValidation.java:
```diff
@@ -0,0 +1,55 @@
+
+    public void validateLengthContentComment(Comment comment) {
+        String content = comment.getContent();
+        if (content.isEmpty()) {
+            throw new DataValidationException("the length of the comment is empty");
+        }
```
Вроде всё же желательно прописывать лог об ошибке перед тем, как выбросить её. Поэтому я обычно текст ошибки в константу выношу, чтобы потом эту константу и в логи записать и в сообщение эксепшена положить. Но тут уже как удобно. Больше не буду об этом писать.

Fagan89 replied:
Логи пока не начинал внедрять. Мы пока просто используем не проверяемые исключения. Когда исключения прога падает как я понял, там уже лог не лог не поможет )) но на 8 спринте как раз будут вроде темы как это правильно делать

tanufa25 commented on src/main/java/faang/school/postservice/validation/comment/CommentValidation.java:
```diff
@@ -0,0 +1,55 @@
+            throw new DataValidationException("the length of the comment is empty");
+        }
+        if (content.length() > MAX_LENGTH_CONTENT) {
+            throw new DataValidationException("the length of the comment is more than 4096 characters");
```
лучше 4096 не явно прописать, а константу в сообщение добавить. А то если вдруг MAX_LENGTH_CONTENT изменится, то в сообщении так и останется 4096.

tanufa25 commented on src/main/java/faang/school/postservice/validation/comment/CommentValidation.java:
```diff
@@ -0,0 +1,55 @@
+
+    public void validateLengthContentComment(Comment comment) {
+        String content = comment.getContent();
+        if (content.isEmpty()) {
```
Возможно, даже эта проверка излишняя, учитываю, что @NotBlank у тебя стоит в CommentCreateDto и в CommentUpdateDto

tanufa25 commented on src/test/java/faang/school/postservice/mapper/comment/MapperCommentTest.java:
```diff
@@ -0,0 +1,74 @@
+//package faang.school.postservice.mapper.comment;
```
Вроде на мапперы и не нужны тесты

Commit:
BJS2-78038
Создал классы (!!!не рабочий код)

Commit:
BJS2-78038
1. Система комментариев
методы: создать, обновить, вернуть все комментарии поста, удалить

2. Покрыл тестами

dextomi commented on src/main/java/faang/school/postservice/controller/comment/CommentController.java:
```diff
@@ -0,0 +1,49 @@
+public class CommentController {
+    private final CommentServiceFacade commentServiceF;
+
+    @PostMapping("/comments")
```
Раз у тебя везде /comments, думаю, это стоит вывести в base url

dextomi commented on src/main/java/faang/school/postservice/controller/comment/CommentController.java:
```diff
@@ -0,0 +1,49 @@
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/v1")
```
/api/v1/comments

dextomi commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,58 @@
+        commentRepository.deleteById(commentId);
+    }
+
+    public Comment getComment(long commentId) {
```
@Transactional(readOnly = true)

Commit:
BJS2-78038
1. скорректировал эндпоинты
2. скорректировал маппер

Commit:
BJS2-78038
1. добавил в контроллер продюсер

Commit:
Merge branch 'hippogriff-master-stream10' into feature/BJS2-78038-system-comment

20cognosce commented on src/main/java/faang/school/postservice/dto/comment/CommentCreateDto.java:
```diff
@@ -0,0 +1,21 @@
+@Builder
+public class CommentCreateDto {
+
+    @NotBlank
```
Применимо только к строкам

20cognosce commented on src/main/java/faang/school/postservice/controller/comment/CommentController.java:
```diff
@@ -0,0 +1,55 @@
+public class CommentController {
+    private final CommentServiceFacade commentServiceF;
+
+    @PostMapping(produces = APPLICATION_JSON_VALUE)
```
не обязательно ставить produces = APPLICATION_JSON_VALUE, итак к json сконвертируется ответ

20cognosce commented on src/main/java/faang/school/postservice/controller/comment/CommentController.java:
```diff
@@ -0,0 +1,55 @@
+        return ResponseEntity.ok(commentDtoResponse);
+    }
+
+    @PutMapping(produces = APPLICATION_JSON_VALUE)
```
Patch

20cognosce commented on src/main/java/faang/school/postservice/service/comment/CommentServiceFacade.java:
```diff
@@ -0,0 +1,44 @@
+
+        return comments.stream()
+                .map(mapperComment::fromEntityToDto)
+                .toList();
```
в маппер

20cognosce commented on src/main/java/faang/school/postservice/validation/comment/CommentValidation.java:
```diff
@@ -0,0 +1,68 @@
+public class CommentValidation {
+    public static final int MAX_LENGTH_CONTENT = 4096;
+
+    private final UserServiceClient userServiceClient;
```
валидатор отсылает запросы во внешний сервис, ну это уже перебор для его зоны ответственности

20cognosce commented on src/main/java/faang/school/postservice/validation/comment/CommentValidation.java:
```diff
@@ -0,0 +1,68 @@
+        }
+    }
+
+    public void validateAuthorExists(Comment comment) {
+        Long authorId = comment.getAuthorId();
+        userServiceClient.getUser(authorId);
+    }
```
в сервис, и не просто делать вызов, а сверять id ответа

20cognosce commented on src/main/java/faang/school/postservice/validation/comment/CommentValidation.java:
```diff
@@ -0,0 +1,68 @@
+        userServiceClient.getUser(authorId);
+    }
+
+    public void validateCommentEqualsUpdateComment(Comment comment, Comment updateComment) {
```
валидация - это проверка на позитивный исход
а у тебя в методе написано что ты валидируешь совпадение, но на самом деле ты ошибку кидаешь - яснее пиши названия методов, checkContentNotEquals(String oldContent, String newContent)

20cognosce commented on src/main/java/faang/school/postservice/validation/comment/CommentValidation.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    public void validatePostExists(Comment comment) {
+        Post post = comment.getPost();
```
Это только внутри транзакции возможно

20cognosce commented on src/main/java/faang/school/postservice/validation/comment/CommentValidation.java:
```diff
@@ -0,0 +1,68 @@
+        if (null == post) {
+            throw new DataValidationException("the comment is not linked to the post");
+        }
+        validatePostExistsById(post.getId());
```
я не понимаю, что ты делаешь
ты из коммента достал пост, откуда он там взялся? Наверно из БД подтянулся. Зачем ты ищешь этот же пост в БД? я не понимаю цель этой проверки

20cognosce commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,58 @@
+    public Comment createComment(Comment comment) {
+        commentValidation.validateLengthContentComment(comment);
+        commentValidation.validateAuthorExists(comment);
+        commentValidation.validatePostExists(comment);
```
Отдели мух от котлет, не нужно в маппере связные сущности, их id, в объекты преобразовывать и сразу в коммент вставлять. 

authorId и postId отдельными параметрами в метод передавай...

20cognosce commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    @Transactional
+    public Comment updateComment(Comment updateComment) {
```
если можно обновить только String content, его и передавай, вместе с commentId, зачем тебе тут автора и пост валидировать (они валидировались при создании), зачем тебе целый объект создавать для update?

20cognosce replied:
Comment updateComment(Long commentId, String newContent)
такая сигнатура

и внутри проверка, что текущий юзер является автором, почему ты разрешаешь редактировать чужие комментарии???

20cognosce commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,58 @@
+
+    @Transactional(readOnly = true)
+    public List<Comment> getAllComments(long postId) {
+        commentValidation.validatePostExistsById(postId);
+        return commentRepository.findAllByPostId(postId);
```
findAllByPostId ничего не вернет или кинет ошибку, валидация бессмысленна

20cognosce commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,58 @@
+
+    @Transactional
+    public void deleteComment(long commentId) {
+        commentValidation.validateCommentExists(commentId);
+        commentRepository.deleteById(commentId);
```
аналогично

и ты бы лучше проверил, что мы свой коммент удаляем, а не других людей без их согласия

Commit:
BJS2-78038
пофиксил замечания

20cognosce commented on src/main/java/faang/school/postservice/controller/comment/CommentController.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    @DeleteMapping("/{commentId}")
+    public ResponseEntity<Long> deleteComment(@PathVariable long commentId,  @RequestParam long authorId) {
```
зачем передавать @RequestParam long authorId?

20cognosce commented on src/main/java/faang/school/postservice/dto/comment/CommentCreateDto.java:
```diff
@@ -0,0 +1,19 @@
+public class CommentCreateDto {
+
+    private Long postId;
+    private Long authorId;
```
и это зачем. инициатор запроса и есть автор
мы что собираемся создавать комменты от чужого имени?

20cognosce commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,74 @@
+
+    @Transactional
+    public Comment createComment(Comment comment) {
+        long authorId = comment.getAuthorId();
+        long postId = comment.getPost().getId();
```
это отдельными параметрами приходит в метод, не в составе сущности, потому что это другие сущности со своим контекстом

автор вообще не должен приходить с фронта, ты его из UserContext достаешь

и пост по id из БД запрашиваешь, а не просто сетишь в маппере

20cognosce commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,74 @@
+        commentValidation.validatePostExists(postId);
+
+        List<Comment> comments = commentRepository.findAllByPostId(postId);
+        if (null == comments) {
+            throw new NullPointerException("List comments cannot be Null");
+        }
```
null вообще не приходит вроде, только пустой список

20cognosce commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,74 @@
+
+    @Transactional(readOnly = true)
+    public List<Comment> getAllComments(long postId) {
+        commentValidation.validatePostExists(postId);
```
какой в этом смысл, если ты итак запрос комментариев делаешь по посту

20cognosce commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,74 @@
+    public void deleteComment(long commentId, long authorId) {
+        commentValidation.validateCommentExists(commentId);
+        Comment comment = getComment(commentId);
+        commentValidation.checkAuthorEqualsUser(authorId, comment.getAuthorId());
```
опять, автор записан в самом комменте, почему ты в метод его принимаешь параметром?
а инициатор запроса - это обязательно не автор, это надо проверить

20cognosce replied:
public void deleteComment(long commentId)
такая сигнатура

20cognosce reviewed (changes requested):
Разберись с DTO, маппером и методами сервиса

Commit:
BJS2-78038
поправил метод удаления коммента

Commit:
BJS2-78038
изменил все методы

Commit:
BJS2-78038
1. доработал тесты
2. убрал лишнии ДТО и поправил маппер

Commit:
BJS2-78038
рефакторинг

Commit:
BJS2-78038
Добавил Retry к UserClient

Commit:
BJS2-78038
fix

Commit:
BJS2-78038
убрал ошибки в контроллере

20cognosce commented on src/main/java/faang/school/postservice/controller/comment/CommentController.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @PatchMapping("/{commentId}")
+    public ResponseEntity<CommentDtoResponse> updateComment(@PathVariable long commentId,
+                                                            @RequestParam String newContent) {
```
в контроллере мог бы оставить DTO, я просил только в сервисе

20cognosce commented on src/main/java/faang/school/postservice/controller/comment/CommentController.java:
```diff
@@ -0,0 +1,49 @@
+
+    @PostMapping("/post/{postId}")
+    public ResponseEntity<CommentDtoResponse> createComment(@PathVariable long postId,
+                                                            @RequestParam String content) {
```
как многострочный комментарий можно будет передавать через query? это тело запроса и передается в DTO

20cognosce commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,83 @@
+    private final PostRepository postRepository;
+
+    @Transactional
+    public Comment createComment(long postId, String content) {
+        long authorId = userContext.getUserId();
+        Post post = postRepository.findById(postId)
+                .orElseThrow(() -> new EntityNotFoundException("the post was not found in the database"));
+        UserDto userDto = userServiceClient.getUser(authorId);
+
+        commentValidation.validateLengthContentComment(content);
+        commentValidation.checkAuthorEqualsUser(authorId, userDto.id());
+
+        Comment comment = Comment.builder()
+                .authorId(authorId)
+                .post(post)
+                .content(content)
+                .build();
+
+        return commentRepository.save(comment);
```
Хорошо

20cognosce reviewed (approved):
В контроллере верни DTO

Commit:
BJS2-78038
вернул Dto, поправил контроллер

Fagan89 commented:
поправил Дто
